persistence/src/main/scala/com/lightbend/lagom/internal/persistence/ReadSideActor.scala

```diff
@@ -0,0 +1,170 @@
+    processor:            () => ReadSideProcessor[Event],
+    eventStreamFactory:   (AggregateEventTag[Event], Offset) => Source[akka.japi.Pair[Event, Offset], NotUsed],
+    clazz:                Class[Event],
+    globalPrepare:        ActorRef,
+    globalPrepareTimeout: FiniteDuration
+  )(implicit mat: Materializer) = {
+    Props(classOf[ReadSideActor[Event]], processor, eventStreamFactory, clazz, globalPrepare, globalPrepareTimeout, mat)
+  }
+
+  /**
+   * Start processing from the given offset
+   */
+  case class Start(offset: Offset)
+}
+
+/**
+ * Read side actor
+ */
+private[lagom] class ReadSideActor[Event <: AggregateEvent[Event]](
+  processorFactory:     () => ReadSideProcessor[Event],
+  eventStreamFactory:   (AggregateEventTag[Event], Offset) => Source[akka.japi.Pair[Event, Offset], NotUsed],
+  clazz:                Class[Event],
+  globalPrepare:        ActorRef,
+  globalPrepareTimeout: FiniteDuration
+)(implicit mat: Materializer) extends Actor with ActorLogging {
+  import ReadSideActor._
+  import akka.pattern.pipe
+  import akka.pattern.ask
+  import context.dispatcher
+
+  private var shutdown: Option[KillSwitch] = None
+
+  override def postStop: Unit = {
+    shutdown.foreach(_.shutdown())
+  }
+
+  def receive = {
+    case EnsureActive(tagName) =>
+
+      val tag = AggregateEventTag.of(clazz, tagName)
+
+      implicit val timeout = Timeout(globalPrepareTimeout)
+
+      globalPrepare ? Prepare pipeTo self
+      context become preparing(tag)
+  }
+
+  def preparing(tag: AggregateEventTag[Event]): Receive = {
+
+    case Done =>
+      val handler = processorFactory().buildHandler()
+      handler.prepare(tag).toScala.map(Start(_)) pipeTo self
+      context become active(handler, tag)
+
+    case Status.Failure(e) =>
+      throw e
+
+  }
+
+  def active(handler: ReadSideProcessor.ReadSideHandler[Event], tag: AggregateEventTag[Event]): Receive = {
+
+    case Start(offset) =>
+
+      val (killSwitch, streamDone) = eventStreamFactory(tag, offset).asScala
+        .viaMat(KillSwitches.single)(Keep.right)
+        .via(handler.handle())
+        .toMat(Sink.ignore)(Keep.both)
+        .run()
+
+      shutdown = Some(killSwitch)
+      streamDone pipeTo self
+
+    case EnsureActive(_) =>
+    // Yes, we are active
+
+    case Status.Failure(e) =>
+      throw e
+
+    case Done =>
+      throw new IllegalStateException("Stream terminated when it shouldn't")
+
+  }
+
+}
+
+private[lagom] object GlobalPrepareReadSideActor {
+
+  /**
+   * Ask the actor if we are prepared yet.
+   *
+   * The actor will send [[akka.Done]] in response.
+   */
+  case object Prepare
+
+  def props(processorFactory: () => ReadSideProcessor[_], prepareTimeout: FiniteDuration) = {
+    Props(classOf[GlobalPrepareReadSideActor], processorFactory, prepareTimeout)
+  }
+}
+
+private[lagom] class GlobalPrepareReadSideActor(processorFactory: () => ReadSideProcessor[_], prepareTimeout: FiniteDuration) extends Actor {
```

Comment:
@patriknw You should check over the logic of this actor, there are three states, the initial state, when it receives a Prepare message it transitions to the preparing state, in that state all subsequent Prepare messages result in the sender being buffered, when it receives a Done message it transitions to the prepare state and tells all the buffered senders that it's prepared.  A failure triggers the actor to crash, and it's managed by a back off supervisor, and when the actor starts it will ask itself to prepare, but put a time limit on it, and then pipe the ask result to itself, so it will crash itself if it exceeds that timeout.